src/GafferScene/CollectScenes.cpp

```diff
@@ -0,0 +1,430 @@
+{
+	if( path.size() == 0 )
+	{
+		return M44f();
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		return inPlug()->transformPlug()->getValue();
+	}
+}
+
+void CollectScenes::hashAttributes( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	if( path.size() == 0 )
+	{
+		SceneProcessor::hashAttributes( path, context, parent, h );
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		h = inPlug()->attributesPlug()->hash();
+	}
+}
+
+IECore::ConstCompoundObjectPtr CollectScenes::computeAttributes( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	if( path.size() == 0 )
+	{
+		return outPlug()->attributesPlug()->defaultValue();
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		return inPlug()->attributesPlug()->getValue();
+	}
+}
+
+void CollectScenes::hashObject( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	if( path.size() == 0 )
+	{
+		SceneProcessor::hashObject( path, context, parent, h );
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		h = inPlug()->objectPlug()->hash();
+	}
+}
+
+IECore::ConstObjectPtr CollectScenes::computeObject( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	if( path.size() == 0 )
+	{
+		return outPlug()->objectPlug()->defaultValue();
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		return inPlug()->objectPlug()->getValue();
+	}
+}
+
+void CollectScenes::hashChildNames( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	if( path.size() == 0 )
+	{
+		SceneProcessor::hashObject( path, context, parent, h );
+		rootNamesPlug()->hash( h );
+	}
+	else
+	{
+		SceneScope sceneScope( context, rootNameVariablePlug()->getValue(), path );
+		h = inPlug()->childNamesPlug()->hash();
+	}
+}
+
+IECore::ConstInternedStringVectorDataPtr CollectScenes::computeChildNames( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	if( path.size() == 0 )
+	{
+		ConstStringVectorDataPtr rootNamesData = rootNamesPlug()->getValue();
+		const vector<string> &rootNames = rootNamesData->readable();
+
+		InternedStringVectorDataPtr childNamesData = new InternedStringVectorData;
+		vector<InternedString> &childNames = childNamesData->writable();
+
+		for( vector<string>::const_iterator it = rootNames.begin(), eIt = rootNames.end(); it != eIt; ++it )
+		{
+			if( it->empty() )
+			{
+				continue;
+			}
+			InternedString childName( *it );
+			if( find( childNames.begin(), childNames.end(), childName ) == childNames.end() )
```

Comment:
Yeah, good idea. I should benchmark this but I suspect things have to get pretty big before a set becomes a win - InternedString comparisons are really fast (single pointer comparison), and they're all stored contiguously in the vector which has much better locality than a set. I think making a generic algorithm is probably going to be problematic because the threshold at which `set<T>` is a win is so dependent on the size and comparison cost of `T`. Since this affects other nodes (and some of them have slightly different requirements too), can we leave this for a separate PR?